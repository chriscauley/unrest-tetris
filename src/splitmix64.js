/*  Written in 2015 by Sebastiano Vigna (vigna@acm.org)

seamlessly translated to JavaScript using BigInt by Elad Karako https://icompile.eladkarako.com/splitmix64-js (January 2020).

https://gist.githubusercontent.com/eladkarako/bf3f8a6017bba9b69185e6e7b4757513/raw/f313292561e8a96b803a3ad31c2e4fd05e90e729/splitmix64.js
*/

const C1 = BigInt(0x9e3779b97f4a7c15) //saves generating BigInt constructor again and again.
const C2 = BigInt(0xbf58476d1ce4e5b9)
const C3 = BigInt(0x94d049bb133111eb)
const F1 = BigInt(30) //shifting
const F2 = BigInt(27)
const F3 = BigInt(31)

export default (seed, MAX_INT = BigInt(Number.MAX_SAFE_INTEGER)) => {
  let x = BigInt(seed)

  const rand = () => {
    x += C1

    let z = x
    z = z ^ ((z >> F1) * C2)
    z = z ^ ((z >> F2) * C3)
    z = z ^ (z >> F3)
    return Number(z % MAX_INT)
  }

  rand.choice = (array) => array[rand() % array.length]
  return rand
}
