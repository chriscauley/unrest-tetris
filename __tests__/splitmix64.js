import splitmix64 from '../src/splitmix64'

test('splitmix64', () => {
  const rand = splitmix64(0)
  const first50 = new Array(50).fill(0).map(() => rand() % 8)
  expect(first50.join(',')).toMatchSnapshot()
})
