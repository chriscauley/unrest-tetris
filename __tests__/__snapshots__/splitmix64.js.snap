// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitmix64 1`] = `"3,1,5,6,6,4,0,3,0,2,0,3,7,2,7,7,1,3,4,2,5,2,5,0,1,1,2,5,5,1,2,3,6,2,2,6,1,4,6,4,4,0,2,2,5,4,3,7,7,3"`;
